core/common/src/main/java/alluxio/master/PortRegistry.java

```diff
@@ -0,0 +1,154 @@
+ * prevent conflicts.
+ *
+ * The default coordination directory is determined by the "user.dir" jvm property. The coordination
+ * directory can be overridden by setting the ALLUXIO_PORT_COORDINATION_DIR environment variable.
+ */
+public final class PortRegistry {
+  private static final String PORT_COORDINATION_DIR_PROPERTY = "ALLUXIO_PORT_COORDINATION_DIR";
+
+  public static final PortRegistry INSTANCE = new PortRegistry();
+
+  // Map from port number to the reservation for that port.
+  private final Map<Integer, Reservation> mReserved = new ConcurrentHashMap<>();
+  private final File mCoordinationDir;
+
+  @VisibleForTesting
+  PortRegistry() {
+    String dir = System.getenv(PORT_COORDINATION_DIR_PROPERTY);
+    if (dir == null) {
+      dir = System.getProperty("user.dir");
+    }
+    mCoordinationDir = new File(dir, ".port_coordination");
+    mCoordinationDir.mkdirs();
+  }
+
+  /**
+   * Reserves a free port so that other tests will not take it.
+   *
+   * @return the free port
+   */
+  public int reservePort() {
+    for (int i = 0; i < 1000; i++) {
+      int port = getFreePort();
+      if (lockPort(port)) {
+        return port;
+      }
+    }
+    throw new RuntimeException("Failed to acquire port");
+  }
+
+  /**
+   * Attempts to lock the given port.
+   *
+   * @param port the port to lock
+   * @return whether the locking succeeded
+   */
+  @VisibleForTesting
+  boolean lockPort(int port) {
+    File portFile = portFile(port);
+    try {
+      FileChannel channel = new RandomAccessFile(portFile, "rw").getChannel();
+      FileLock lock = channel.tryLock();
+      if (lock == null) {
+        channel.close();
+        return false;
+      }
+      mReserved.put(port, new Reservation(portFile, lock));
+      return true;
+    } catch (IOException | OverlappingFileLockException e) {
+      return false;
+    }
+  }
+
+  /**
+   * @param port the port to release
+   */
+  public void release(int port) {
+    Reservation r = mReserved.remove(port);
+    if (r != null) {
+      // If delete fails, we may leave a file behind. However, the file will be unlocked, so another
+      // process can still take the port.
+      r.mFile.delete();
+      try {
+        r.mLock.release();
+        r.mLock.channel().close();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  /**
+   * Clears the registry.
+   */
+  public void clear() {
+    new HashSet<>(mReserved.keySet()).forEach(this::release);
+  }
+
+  @VisibleForTesting
+  File portFile(int port) {
+    return new File(mCoordinationDir, Integer.toString(port));
+  }
+
+  /**
+   * @return a port that is currently free. This does not reserve the port, so the port may be taken
+   *         by the time this method returns.
+   */
+  public static int getFreePort() {
+    int port;
+    try {
+      ServerSocket socket = new ServerSocket(0);
+      socket.setReuseAddress(true);
```

Comment:
We had a short discussion prior - but what exactly does setting reuseAddress here do for us?